File: 03_JavaScript/02_Stack/Stack.js
console.clear();

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }

    toString() {
        return JSON.stringify(this, null, 2);
    }
}

class Stack {
    #length;
    #head;
    #tail;

    constructor() {
        this.#length = 0;
        this.#head = null;
        this.#tail = null;
    }

    get length() {
        return this.#length;
    }

    get getHead() {
        return this.#head;
    }

    get getTail() {
        return this.#tail;
    }

    /**
     * @param {any} data 
     */
    push(data) {
        const node = new Node(data);
        let head = this.getHead;
        this.#length++;

        if (head === null) {
            this.#tail = node;
        }

        node.next = head;
        this.#head = node;
    }

    /**
     * @returns {Node}
     */
    pop() {
        let lastNode = null;
        const head = this.getHead;

        if (head === null) {
            return lastNode;
        }

        this.#length--;

        if (head.next === null) {
            lastNode = head;
            
            this.clear()
        }

        if (head !== null) {
            const remainNode = head.next;

            head.next = null;
            lastNode = head;

            this.#head = remainNode;
        }

        return lastNode;
    }

    
    iterator() {
        let iterate = this.getHead;
        
        return {
            next: () => {
                const node = iterate;
                
                iterate = node !== null ? iterate.next : this.getHead;
                
                return {
                    done: node === null,
                    data: node ? node.data : null
                }
            }
        }
    }
    
    clear() {
        this.#length = 0;
        this.#head = null;
        this.#tail = null;
    }
    
    toString() {
        return JSON.stringify({
            length: this.length,
            head: this.getHead,
            tail: this.getTail
        }, null, 2);
    }
}

const stack = new Stack();

stack.push(1);
stack.push(2);
stack.push(3);
stack.push(4);
stack.push(5);
stack.push(6);

console.log(stack.toString())

const lastNode = stack.pop();
const secondLastNode = stack.pop();
const thirdLastNode = stack.pop();

console.log(lastNode, secondLastNode, thirdLastNode);

console.log(stack.toString())

const iterate = stack.iterator()
console.log(iterate.next())
console.log(iterate.next())
console.log(iterate.next())
console.log(iterate.next())
console.log(iterate.next())
console.log(iterate.next())
console.log(iterate.next())
console.log(iterate.next())

stack.clear();
console.log(stack.toString());

File: Java/src/Stack/Stack.java
package Stack;

public class Stack {
	private int length;
	private StackNode head;
	private StackNode tail;

	public Stack() {
		this.length = 0;
		this.head = null;
		this.tail = null;
	}

	public int getLength() {
		return this.length;
	}

	public StackNode getHead() {
		return this.head;
	}

	public StackNode getTail() {
		return this.tail;
	}

	public void push(int data) {
		StackNode node = new StackNode(data);
		StackNode head = this.head;

		if (head != null) {
			while (head.next != null) {
				head = head.next;
			}

			head.next = node;
		} else this.head = node;

		this.tail = node;
		this.length++;
	}

	public StackNode pop() {
		StackNode lastNode = this.tail;
		StackNode head = this.head;

		if (head == null) {
			return null;
		}

		this.length--;

		if (head.next != null) {
			while (head.next.next != null) {
				head = head.next;
			}

			head.next = null;
		} else this.clear();


		this.tail = head;
		return lastNode;
	}

	public void clear() {
		this.length = 0;
		this.head = null;
		this.tail = null;
	}

	@Override
	public String toString() {
		return "Stack {" + '\n' +
		" length=" + length + ",\n" +
		" head=" + head + ",\n" +
		" tail=" + tail + '\n' +
		'}';
	}
}
